pkg/querier/tripperware/query_attribute_matcher.go

```diff
@@ -0,0 +1,237 @@
+		return false
+	}
+
+	return true
+}
+
+func matchAttributeForMetadataQuery(attribute validation.QueryAttribute, r *http.Request, now time.Time) bool {
+	if err := r.ParseForm(); err != nil {
+		return false
+	}
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.CompiledRegex != nil {
+		atLeastOneMatched := false
+		for _, matcher := range r.Form["match[]"] {
+			if attribute.CompiledRegex.MatchString(matcher) {
+				atLeastOneMatched = true
+				break
+			}
+		}
+		if !atLeastOneMatched {
+			return false
+		}
+	}
+
+	startTime, _ := util.ParseTime(r.FormValue("start"))
+	endTime, _ := util.ParseTime(r.FormValue("end"))
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, startTime, endTime) {
+		return false
+	}
+
+	if !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, startTime, endTime) {
+		return false
+	}
+
+	if attribute.UserAgentRegex != "" && attribute.UserAgentRegex != ".*" && attribute.CompiledUserAgentRegex != nil {
+		if !attribute.CompiledUserAgentRegex.MatchString(r.Header.Get("User-Agent")) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isWithinTimeAttributes(timeWindow validation.TimeWindow, now time.Time, startTime, endTime int64) bool {
+	if timeWindow.Start == 0 && timeWindow.End == 0 {
+		return true
+	}
+
+	if timeWindow.Start != 0 {
+		startTimeThreshold := now.Add(-1 * time.Duration(timeWindow.Start).Abs()).Add(-1 * time.Minute).Truncate(time.Minute).UnixMilli()
+		if startTime == 0 || startTime < startTimeThreshold {
+			return false
+		}
+	}
+
+	if timeWindow.End != 0 {
+		endTimeThreshold := now.Add(-1 * time.Duration(timeWindow.End).Abs()).Add(1 * time.Minute).Truncate(time.Minute).UnixMilli()
+		if endTime == 0 || endTime > endTimeThreshold {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isWithinTimeRangeAttribute(limit validation.TimeRangeLimit, startTime, endTime int64) bool {
+	if limit.Min == 0 && limit.Max == 0 {
+		return true
+	}
+
+	if startTime == 0 || endTime == 0 {
+		return false
+	}
+
+	timeRangeInMillis := endTime - startTime
+
+	if limit.Min != 0 && time.Duration(limit.Min).Milliseconds() > timeRangeInMillis {
+		return false
+	}
+	if limit.Max != 0 && time.Duration(limit.Max).Milliseconds() < timeRangeInMillis {
+		return false
+	}
+
+	return true
+}
+
+func isWithinQueryStepLimit(queryStepLimit validation.QueryStepLimit, r *http.Request, expr parser.Expr) bool {
+	if queryStepLimit.Min == 0 && queryStepLimit.Max == 0 {
+		return true
+	}
+	var stepLimitChecked bool
+
+	if step, err := util.ParseDurationMs(r.FormValue("step")); err == nil {
+		if queryStepLimit.Min != 0 && time.Duration(queryStepLimit.Min).Milliseconds() > step {
+			return false
+		}
+		if queryStepLimit.Max != 0 && time.Duration(queryStepLimit.Max).Milliseconds() < step {
+			return false
+		}
+		stepLimitChecked = true
+	}
+
+	subQueryStepWithinLimit := true
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
```

Comment:
I was thinking of the case where user doesn't do regex, and just want's to limit some query which has big time range with very tiny steps. In this case query_attribute will have only time_range_limit and query_step_limit properties. This way could be more clear for user. And another reason I included subquery steps as well is because users expectation may be that when there step limit it should be applied to both steps.

Reply:
> And another reason I included subquery steps as well is because users expectation may be that when there step limit it should be applied to both steps

I am not sure if it is true. Subquery step is used differently and Prometheus itself doesn't even have a limit for subquery.

I can see usecases where users want different limits. Like min allowed step is 1min but subquery step can be low down to 30s.